### PR TITLE
Implement a plugin system in C++ and integrate `ros2_control`

### LINKED ISSUE #210: Integrate `ros2_control`
The [ros2_control](https://discourse.ros.org/t/announcing-ros2-control-for-foxy/18274) package has been released recently and it would useful to integrate it.

Once we have the [ros2_control](https://github.com/ros-controls/ros2_control) integrated we can get rid of:
- `webots_differential_drive_node.py` in favor of [diff_drive_controller](https://github.com/ros-controls/ros2_controllers/tree/master/diff_drive_controller)
- `joint_state_publisher.py` in favor of [joint_state_controller](https://github.com/ros-controls/ros2_controllers/tree/master/joint_state_controller)
- `trajectory_follower.py` in favor of [joint_trajectory_controller](https://github.com/ros-controls/ros2_controllers/tree/master/joint_trajectory_controller)
- `webots_robotic_arm_node.py` will be useless as well.

Each controller coming from the `ros2_controllers` package is better designed + we will get access to [more controllers](https://github.com/ros-controls/ros2_controllers) (there will be even more as they have recently started porting them from ROS 1). Other benefits are:
- the ability to create a custom controller or use an existing custom controller,
- use `ros2controlcli`,
- replace controllers at runtime,
- manage ownership,
- and other perks coming with the [ros2_control](http://www.theoj.org/joss-papers/joss.00456/10.21105.joss.00456.pdf).

The problem is that `ros2_control` provides only a C++ interface. Therefore, there are two ways to integrate the package:
1. Translate `webots_ros2_core/webots_node.py` and `webots_ros2_core/devices` to C++, and create Python bindings for `webots_node.py` so `WebotsNode` can be extended in Python as well.
2. Create a standalone `ros2_control` based node that will exchange metadata about joints with the Python driver. It means that the `read/write` functions would interact with `WebotsNode` through custom ROS2 messages.

Resources:
- [ros2_control docs](https://ros-controls.github.io/control.ros.org/getting_started.html#compiling)
- [ros_control in general](https://fjp.at/posts/ros/ros-control/)

### PR DESCRIPTION
**Pull request by lukicdarkoo**
This PR should give a base for a system that supports creating custom plugins in C++ and Python, and it should resolve #210.

Why switching to C++?
- seamless integration with `ros2_control`,
- allows plugins to written in C++,
- performance

Why not?
- managing the Webots controller dependency is cumbersome,
- the binary will be only compatible with a single Webots version (can support multiple by it is requires extra work),

**Architecture**
https://docs.google.com/drawings/d/15oj_hYvYZ5Sn8JmSoqBsnzu-DPusRuWWLy8nW5L-C3k/edit?usp=sharing

**Implementation**
- Pass `WebotsNode` to the plugin. In Python case:
  - `WebotsNode` will contain bindings for some methods from `rlcpp` (like `create_service`, `create_topic`, `create_timer`) and 
  -  some `webots_ros2` specific methods (like `name`).
  - However, it will not contains bindings for `Robot/Supervisor/Driver`, but C++ reference will be passed to the Webots Python library and their bindings will be resolved. 

**TODO**
- [x] Implement a plugin system for C++
- [x] Utilize the `<webots>` tag for plugins
- Add most common devices (devices implemented in the previous version)
  - [x] Camera
  - [x] GPS
  - [x] LiDAR
  - [x] Range Finder
  - [x] IMU (has to be a plugin as it is composed of three Webots devices)
  - [x] DistanceSensor
  - [x] LED
  - [x] LightSensor
- Integrate `ros2_control`
  - [x] Add a `webots_ros2_control` package
  - [x] Allow the `hardware_interface::ResourceManager` to load multiple system interfaces
  - [x] Spawn a `controller_manager::ControllerManager` node
  - [x] Veirfy the behaviour with `diffdriver_controller` and `joint_state_publisher` 
- Convert the TurtleBot3 Burger example (to serve as a test for the new system)
  - [x] Use the new C++ interface
  - [x] Use the `webots_ros2_control` 
  - [x] Test SLAM and navigation
- [x] Configure the CI to build the package
  - Unfortunately, I believe we don't have access to the internet from ROS Build farm, so I am not sure if we can download it later. Maybe the easiest way to resolve Webots controller dependency is to manually copy all necessary files into this repository. The controller size is ~1MB (.so files) + 200KB (.h/hpp files). But note also that these files have to be copied for each OS and Webots version.
- [x] Publish simulation time
- [x] Changelog
- [x] Package descriptions
- [x] Rename `webots_ros2_cpp` to `webots_ros2_bridge` (and `webots_ros2` namespace to `webots_ros2_bridge`)
- [x] Move `webots_ros2_cpp/webots/upstream` to `webots_ros2_cpp/webots`
- Fix
  - [x] There are some strange performance issues that started happening
  - [x] We should be able to avoid the `package` attribute in the plugins

**TODO 2** (other PRs)
- Convert all examples to use the new interface
- Mark the old interface as a deprecated
- Implement Python plugin system
- Integrate `urdf2webots` to dynamically populate Robot's children parameter (https://discourse.ros.org/t/urdf-with-webots/19665)
  - We have to create a ROS 2 Webots spawner node that takes a URDF, converts it to VRML, and passes it (as a topic or parameter) to the `WebotsNode`.

**Notes**
```bash
ros2 launch webots_ros2_turtlebot robot_launch.py
```

### CODE REVIEW

**Comment by lukicdarkoo** [comment]
@cyberbotics/maintainers @renan028 There are small improvements and fixes I have to do, but the PR is ready for the first round of reviews.

**Review by omichel — COMMENTED** [review]
It looks very good to me so far.
Regarding the namespace, I have no better idea than "webots_ros2_interface"...

**Comment by renan028** [comment]
wow! Very nice that you manage to push that forward.
It seems very good. I will try to review it

**Comment by lukicdarkoo** [comment]
We can rename `webots_ros2_interface` to `webots_ros2_driver` or `webots_ros2_bridge` (like [here](https://github.com/lgsvl/ros2-lgsvl-bridge))

I covered all points, so it is ready for full review.

**Comment by omichel** [comment]
I tend to prefer `webots_ros2_driver` as `interface` is long and too generic and `bridge` may have some slow performance connotation.

**Review by omichel — APPROVED** [review]
It looks good to me, but I would wait for the review of @renan028 before merging it.

File: webots_ros2_control/src/Ros2ControlSystem.cpp
Comment (renan028) [inline webots_ros2_control/src/Ros2ControlSystem.cpp:60]: else {...}

maybe warning and skip?

File: webots_ros2_driver/include/webots_ros2_driver/WebotsNode.hpp
Comment (renan028) [inline webots_ros2_driver/include/webots_ros2_driver/WebotsNode.hpp:49]: `unordered_map`

File: webots_ros2_driver/include/webots_ros2_driver/plugins/Ros2SensorPlugin.hpp
Comment (renan028) [inline webots_ros2_driver/include/webots_ros2_driver/plugins/Ros2SensorPlugin.hpp:37]: maybe add these pure virtual methods?
```
enable()=0
disable()=0
preStep()=0
```

File: webots_ros2_driver/src/WebotsNode.cpp
Comment (renan028) [inline webots_ros2_driver/src/WebotsNode.cpp:63]: `unordered_map`

Comment (renan028) [inline webots_ros2_driver/src/WebotsNode.cpp:63]: this method can be const

File: webots_ros2_driver/src/WebotsNode.cpp
Comment (renan028) [inline webots_ros2_driver/src/WebotsNode.cpp:80]: this method can be const

File: webots_ros2_driver/src/WebotsNode.cpp
Comment (renan028) [inline webots_ros2_driver/src/WebotsNode.cpp:63]: I think this method can be const

File: webots_ros2_driver/include/webots_ros2_driver/plugins/Ros2SensorPlugin.hpp
Comment (lukicdarkoo) [reply webots_ros2_driver/include/webots_ros2_driver/plugins/Ros2SensorPlugin.hpp:38]: We don't have enable/disable methods and `preStep()` should not be overridden in general. It should be protected though.

**Comment by lukicdarkoo** [comment]
@renan028 Thank you for the review, I went through your comments, let me know if there anything else that should be fixed/improved

**Review by renan028 — APPROVED** [review]

**Comment by renan028** [comment]
It looks good to me

**Comment by Timple** [comment]
Awesome work! I was hoping to test this new feature. I can't seem to find a compatible version set. Would you mind sharing the versions you tested with?

ros2: foxy/galactic/rolling?
ros2_control: master?
ros2_controllers: master?
webots_ros2: master?
webots: R2021a? Using apt?

**Comment by lukicdarkoo** [comment]
@Timple, thank you for pointing out to missing information, we still have to convert all examples and document everything. Here:

- Webots: R2021b (but R2021a should also be fine).
- ROS 2: Foxy. `ros2_control` is only available for Foxy, but I am looking forward to https://github.com/ros-controls/ros2_control/issues/421.
- `ros2_control/ros2_controllers`: Those are compiled from master. Strangely, the apt installation didn't work on my development machine but it works in CI.

> edit: to be specific, some combinations compile fine, but I cannot seem to get a `/cmd_vel` topic to publish to using `ros2 launch webots_ros2_turtlebot robot_launch.py`

You have to activate the controllers (I am not sure if we can avoid it, maybe I should ask the `ros2_control` team), for example:
```bash
ros2 control load_start_controller diffdrive_controller

# Verify
ros2 control list_controllers
```

**Comment by Timple** [comment]
Yes, that seems to be working!

Although the master branch of `ros_controllers` doesn't seem to compile for `gripper_controllers` and `joint_trajectory_controller`. But those where not required here.

Also works for Galactic!

---
> `ros2 control load_start_controller diffdrive_controller`

Might be nice to add that to the Notes in your first post. Just for the folks still getting used to `ros(2)_control`.

**Comment by Timple** [comment]
Should `webots_ros2_turtlebot` also have an `<exec_depend>` on `diff_drive_controller`?

This doesn't get compiled now, but without that package the robot won't do anything.

**Comment by lukicdarkoo** [comment]
> Should `webots_ros2_turtlebot` also have an `<exec_depend>` on `diff_drive_controller`?

You are right, here is the PR https://github.com/cyberbotics/webots_ros2/pull/228

> This doesn't get compiled now

What do you mean? It used to compile and now it doesn't?

**Comment by Timple** [comment]
No, as it wasn't specified as dependency, colcon doesn't see the need to compile it. (When using packages-up-to)

File: webots_ros2_turtlebot/launch/robot_launch.py
Comment (Timple) [inline webots_ros2_turtlebot/launch/robot_launch.py:28]: Sorry for the review in hindsight. But this also introduces a dependency on `webots_ros2_core`

**Comment by Timple** [comment]
> Also works for Galactic!

Sorry, revoking this statement. Was mixing workspaces and still using foxy.

For galactic I get: 
```
[driver-3] [INFO] [1623129717.776449976] [controller_manager]: Loading controller 'diffdrive_controller'
[driver-3] Exception thrown during init stage with message: expected [double_array] got [not set] 
[driver-3] [ERROR] [1623129717.783781570] [controller_manager]: Could not initialize the controller named 'diffdrive_controller'
```

Not sure where to start debugging. Still new to ROS2 so backtraces are out of my comfort zone still :slightly_smiling_face:

**Comment by lukicdarkoo** [comment]
@Timple, Try compiling it in debug mode:
```bash
colcon build --cmake-args -DCMAKE_BUILD_TYPE=RelWithDebInfo
```
Then, open Webots and run the driver with GDB:
```bash
gdb install/webots_ros2_driver/lib/webots_ros2_driver/driver
```
It should let you backtrace the exception with the `bt` command.

**Comment by Timple** [comment]
That's not that easy because of remappings and parameters right?

Also the exception is caught in the code by a `try/catch` construction. So the node does not fail.

I simply commented the `try/catch` and got this:
```
[driver-3] [INFO] [1623137736.407722715] [controller_manager]: Loading controller 'diffdrive_controller'
[driver-3] terminate called after throwing an instance of 'rclcpp::ParameterTypeException'
[driver-3]   what():  expected [double_array] got [not set]
[ERROR] [driver-3]: process has died [pid 1399160, exit code -6, cmd '/home/tim/ros/jax-galactic/system/install/webots_ros2_driver/lib/webots_ros2_driver/driver --ros-args --params-file /tmp/launch_params_jg9kofxc --params-file /home/tim/ros/jax-galactic/system/install/webots_ros2_turtlebot/share/webots_ros2_turtlebot/resource/ros2control.yml -r /diffdrive_controller/cmd_vel_unstamped:=/cmd_vel'].
```

Note that during compilation I also had these warnings which seem related:
<details>
  <summary>Click to expand!</summary>
  
  ```
Starting >>> diff_drive_controller
--- stderr: diff_drive_controller                                
/home/tim/ros/jax-galactic/system/src/diff_drive_controller/src/diff_drive_controller.cpp: In member function ‘virtual controller_interface::return_type diff_drive_controller::DiffDriveController::init(const string&)’:
/home/tim/ros/jax-galactic/system/src/diff_drive_controller/src/diff_drive_controller.cpp:64:77: warning: converting to ‘const ParameterDescriptor’ {aka ‘const rcl_interfaces::msg::ParameterDescriptor_<std::allocator<void> >’} from initializer list would use explicit constructor ‘rcl_interfaces::msg::ParameterDescriptor_<ContainerAllocator>::ParameterDescriptor_(rosidl_runtime_cpp::MessageInitialization) [with ContainerAllocator = std::allocator<void>]’
   64 |     node->declare_parameter<std::vector<std::string>>("left_wheel_names", {});
      |                                                                             ^
/home/tim/ros/jax-galactic/system/src/diff_drive_controller/src/diff_drive_controller.cpp:64:77: note: in C++11 and above a default constructor can be explicit
/home/tim/ros/jax-galactic/system/src/diff_drive_controller/src/diff_drive_controller.cpp:65:78: warning: converting to ‘const ParameterDescriptor’ {aka ‘const rcl_interfaces::msg::ParameterDescriptor_<std::allocator<void> >’} from initializer list would use explicit constructor ‘rcl_interfaces::msg::ParameterDescriptor_<ContainerAllocator>::ParameterDescriptor_(rosidl_runtime_cpp::MessageInitialization) [with ContainerAllocator = std::allocator<void>]’
   65 |     node->declare_parameter<std::vector<std::string>>("right_wheel_names", {});
      |                                                                              ^
/home/tim/ros/jax-galactic/system/src/diff_drive_controller/src/diff_drive_controller.cpp:65:78: note: in C++11 and above a default constructor can be explicit
/home/tim/ros/jax-galactic/system/src/diff_drive_controller/src/diff_drive_controller.cpp:82:80: warning: converting to ‘const ParameterDescriptor’ {aka ‘const rcl_interfaces::msg::ParameterDescriptor_<std::allocator<void> >’} from initializer list would use explicit constructor ‘rcl_interfaces::msg::ParameterDescriptor_<ContainerAllocator>::ParameterDescriptor_(rosidl_runtime_cpp::MessageInitialization) [with ContainerAllocator = std::allocator<void>]’
   82 |     node->declare_parameter<std::vector<double>>("pose_covariance_diagonal", {});
      |                                                                                ^
/home/tim/ros/jax-galactic/system/src/diff_drive_controller/src/diff_drive_controller.cpp:82:80: note: in C++11 and above a default constructor can be explicit
/home/tim/ros/jax-galactic/system/src/diff_drive_controller/src/diff_drive_controller.cpp:83:81: warning: converting to ‘const ParameterDescriptor’ {aka ‘const rcl_interfaces::msg::ParameterDescriptor_<std::allocator<void> >’} from initializer list would use explicit constructor ‘rcl_interfaces::msg::ParameterDescriptor_<ContainerAllocator>::ParameterDescriptor_(rosidl_runtime_cpp::MessageInitialization) [with ContainerAllocator = std::allocator<void>]’
   83 |     node->declare_parameter<std::vector<double>>("twist_covariance_diagonal", {});
      |                                                                                 ^
/home/tim/ros/jax-galactic/system/src/diff_drive_controller/src/diff_drive_controller.cpp:83:81: note: in C++11 and above a default constructor can be explicit
---
Finished <<< diff_drive_controller [13.8s]

  ```
</details>

**Comment by lukicdarkoo** [comment]
> That's not that easy because of remappings and parameters right?

You don't necessarily need them, some partial test should be fine. However, if you need them then you can, you can pass them as arguments:
https://docs.ros.org/en/foxy/Guides/Node-arguments.html#example

For example:
```bash
gdb --args install/webots_ros2_driver/lib/webots_ros2_driver/driver --ros-args --params-file install/webots_ros2_turtlebot/share/webots_ros2_turtlebot/resource/ros2control.yml -r /diffdrive_controller/cmd_vel_unstamped:=/cmd_vel
```

It is possible that the `diff_drive_controller` is not compatible with Galactic. Galactic is released just recently, so there are a lot of moving parts to be checked. 

> PS. sorry for hijacking this thread, let me know if this deserves an issue.

No problem, it is relevant to the PR. But go ahead with opening an issue.

**Comment by Timple** [comment]
I found the culprit already: https://github.com/ros-controls/ros2_controllers/pull/199

Probably to do with the c++14 to c++17 transition.

**Comment by Timple** [comment]
So this is all verified to be working at the moment!

Would be great to have a new bloom-release :pray: :grin:

**Comment by lukicdarkoo** [comment]
@Timple What is your experience with describing the ROS interface and plugins in the `<webots>` tag? How does it differ (from a user point of view) from the Python interface we used before?

**Comment by Timple** [comment]
I would say limited.

In ROS1 we used the python interface. Which was basically writing your own bridge, so a bit of a hassle. Did leave a lot of room for custom implementations though.

For a new ROS2 project/robot I was diving into how to setup or low-level control of the robot properly. And `ros2_control` seems like the way to go. For this project we chose webots as simulator. So this led me straight to this PR :slightly_smiling_face:

**Comment by lukicdarkoo** [comment]
> I would say limited.
> 
> In ROS1 we used the python interface. Which was basically writing your own bridge, so a bit of a hassle. Did leave a lot of room for custom implementations though.
> 
> For a new ROS2 project/robot I was diving into how to setup or low-level control of the robot properly. And `ros2_control` seems like the way to go. For this project we chose webots as simulator. So this led me straight to this PR

For ROS 2 we started by creating a Python bridge that creates topics/services for the most common devices, but you can also simply inherit it to extend its functionalities. However, we noticed some performance issues, and integration with e.g. ros2_control wasn't possible. Therefore, we created`webots_ros2_driver` as a prototype that solves the mentioned issues + you can still extend it using plugins (e.g. `webots_ros2_control` is a plugin). In the future, we plan to allow plugins to be written in Python as well.

If you have ideas on how to improve the interface you can make an issue or you can write to our Discord channel (or me directly there):
https://discordapp.com/invite/nTWbN9m

**Pull request merged**
